src/Order.sol

```diff
@@ -315,51 +400,45 @@ contract Order {
     }
 
     /// @dev Marks the order as sucessfully completed, and transfers the tokens.
-    function confirm(uint32 index)
+    function confirm(address taker_, uint32 index)
         public
         virtual
-        onlyState(msg.sender, index, State.Committed)
+        onlyState(taker_, index, State.Committed)
     {
+        Offer memory offer = offers[taker_][index];
+
+        // Only buyer can confirm before timeout
+        if (block.timestamp < timeout + offer.acceptedAt) {
+            if (msg.sender != buyer(taker_)) {
+                revert MustBeBuyer();
+            }
+        }
+
         // Close the offer
-        Offer memory offer = offers[msg.sender][index];
-        offers[msg.sender][index] = Offer(
+        offers[taker_][index] = Offer(
             State.Closed,
             0,
             0,
+            0,
             '',
             uint64(block.timestamp),
             false,
             false,
             0
         );
 
-        uint256 total = offer.pricePerUnit * offer.quantity;
+        uint256 total = (offer.pricePerUnit + offer.sellerStakePerUnit) * offer.quantity;
         uint256 toOrderBook = (total * IOrderBook(orderBook).fee()) /
             ONE_MILLION;
         uint256 toSeller = total - toOrderBook;
+        uint256 toBuyer = buyerStakePerUnit(offer) * offer.quantity;
```

Comment:
Here the order book fee is now applied to the seller stake as well - is that intentional? I don't recall any discussion on that particular change.